File: src/scala/UnionTypes.scala
//: src/scala/UnionTypes.scala
// Union types aka sum types
package uniontypes
import disjointtypes.*

def nourish(x: Person | Robot) = x match
    case p: Person => p.consume()
    case r: Robot  => r.charge()

@main def main() =
    val list: List[Person | Robot] = List(Person(), Robot())
    list.foreach(nourish)

File: src/python/union_types.py
#: src/python/union_types.py
from disjoint_types import Person, Robot

def nourish(x: Person | Robot):
    match x:  # No exhaustiveness checking (yet)
        case Person(): x.consume()
        case Robot(): x.charge()

if __name__ == '__main__':
    for subject in [Person(), Robot()]:
        nourish(subject)
    # nourish("")  # Type-check error
